File: xue_hua_media_compression_android/android/src/main/kotlin/com/xuehua/xue_hua_media_compression/XueHuaMediaCompressionPlugin.kt
package com.xuehua.xue_hua_media_compression

import android.content.Context
import android.os.Handler
import android.os.Looper
import androidx.media3.common.util.UnstableApi
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.BinaryMessenger
import java.util.concurrent.Executors

/**
 * Android implementation: ImageDecoder/Bitmap/HeifWriter + Media3 Transformer.
 *
 * Android 实现：图片走 ImageDecoder/Bitmap/HeifWriter，视频走 Media3 Transformer。
 */
class XueHuaMediaCompressionPlugin : FlutterPlugin, MediaCompressionHostApi {
  private lateinit var context: Context
  private lateinit var messenger: BinaryMessenger
  private val mainHandler = Handler(Looper.getMainLooper())
  private val executor = Executors.newCachedThreadPool()

  private val jobs = mutableMapOf<Long, Job>()
  private var nextId = 1L

  override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    context = binding.applicationContext
    messenger = binding.binaryMessenger
    MediaCompressionHostApi.setUp(messenger, this)
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    MediaCompressionHostApi.setUp(binding.binaryMessenger, null)
    jobs.values.forEach { it.dispose() }
    jobs.clear()
    executor.shutdownNow()
  }

  override fun createJob(): Long {
    val id = nextId++
    jobs[id] = Job(id, JobEventStream(messenger, id))
    return id
  }

  override fun queryImageCapabilities(): ImageCapabilitiesMsg {
    return ImageCompressor.queryCapabilities()
  }

  override fun startImageCompress(
      id: Long,
      source: SourceMsg,
      destination: DestinationMsg,
      options: ImageOptionsMsg,
      callback: (Result<Unit>) -> Unit,
  ) {
    val job =
        jobs[id]
            ?: return callback(
                Result.failure(FlutterError("instanceNotFound", "No job $id", null)))
    if (job.started) {
      return callback(
          Result.failure(FlutterError("invalidState", "Job $id already started", null)))
    }
    job.started = true
    callback(Result.success(Unit))
    executor.execute {
      try {
        if (job.cancelled) {
          job.events.sendError("cancelled", "Cancelled", null)
          return@execute
        }
        job.events.sendProgress(0.05)
        val result = ImageCompressor.compress(context, source, destination, options)
        if (job.cancelled) {
          job.events.sendError("cancelled", "Cancelled", null)
          return@execute
        }
        job.events.sendProgress(1.0)
        job.events.sendCompleted(result)
      } catch (error: FlutterError) {
        job.events.sendError(error.code, error.message ?: error.code, error.details?.toString())
      } catch (error: Exception) {
        job.events.sendError("encode", error.message ?: "Image compress failed", error.toString())
      }
    }
  }

  override fun queryVideoCapabilities(): VideoCapabilitiesMsg {
    return VideoCompressor.queryCapabilities()
  }

  @UnstableApi
  override fun startVideoCompress(
      id: Long,
      inputPath: String,
      outputPath: String,
      options: VideoOptionsMsg,
      callback: (Result<Unit>) -> Unit,
  ) {
    val job =
        jobs[id]
            ?: return callback(
                Result.failure(FlutterError("instanceNotFound", "No job $id", null)))
    if (job.started) {
      return callback(
          Result.failure(FlutterError("invalidState", "Job $id already started", null)))
    }
    job.started = true
    try {
      val compressor = VideoCompressor(context, job.events)
      job.video = compressor
      compressor.start(inputPath, outputPath, options)
      callback(Result.success(Unit))
    } catch (error: FlutterError) {
      callback(Result.failure(error))
    } catch (error: Exception) {
      callback(
          Result.failure(
              FlutterError("encode", error.message ?: "Video start failed", error.toString())))
    }
  }

  override fun cancelJob(id: Long) {
    val job = jobs[id] ?: throw FlutterError("instanceNotFound", "No job $id", null)
    job.cancelled = true
    job.video?.cancel()
  }

  override fun disposeJob(id: Long) {
    jobs.remove(id)?.dispose()
  }

  private class Job(val id: Long, val events: JobEventStream) {
    var started = false
    @Volatile var cancelled = false
    var video: VideoCompressor? = null

    fun dispose() {
      cancelled = true
      video?.cancel()
      events.dispose()
    }
  }
}

File: packages/xue_hua_media_compression_darwin/darwin/xue_hua_media_compression_darwin/Sources/xue_hua_media_compression_darwin/XueHuaMediaCompressionPlugin.swift
#if os(iOS)
  import Flutter
#elseif os(macOS)
  import FlutterMacOS
#endif
import Foundation

/// iOS / macOS entry point. Registers the Pigeon HostApi.
///
/// iOS / macOS 入口：注册 Pigeon HostApi。
public class XueHuaMediaCompressionPlugin: NSObject, FlutterPlugin, MediaCompressionHostApi {
  public static func register(with registrar: FlutterPluginRegistrar) {
    #if os(macOS)
      let messenger = registrar.messenger
    #else
      let messenger = registrar.messenger()
    #endif
    let plugin = XueHuaMediaCompressionPlugin(messenger: messenger)
    MediaCompressionHostApiSetup.setUp(binaryMessenger: messenger, api: plugin)
    let retain =
      FlutterMethodChannel(
        name: "xue_hua_media_compression/retain", binaryMessenger: messenger)
    registrar.addMethodCallDelegate(plugin, channel: retain)
  }

  private let messenger: FlutterBinaryMessenger
  private var jobs: [Int64: Job] = [:]
  private var nextId: Int64 = 1

  init(messenger: FlutterBinaryMessenger) {
    self.messenger = messenger
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    result(FlutterMethodNotImplemented)
  }

  func createJob() throws -> Int64 {
    let id = nextId
    nextId += 1
    jobs[id] = Job(id: id, events: JobEventStream(messenger: messenger, id: id))
    return id
  }

  func queryImageCapabilities() throws -> ImageCapabilitiesMsg {
    ImageCompressor.queryCapabilities()
  }

  func startImageCompress(
    id: Int64,
    source: SourceMsg,
    destination: DestinationMsg,
    options: ImageOptionsMsg,
    completion: @escaping (Result<Void, Error>) -> Void
  ) {
    guard let job = jobs[id] else {
      completion(
        .failure(PigeonError(code: "instanceNotFound", message: "No job \(id)", details: nil)))
      return
    }
    if job.started {
      completion(
        .failure(PigeonError(code: "invalidState", message: "Job \(id) already started", details: nil)))
      return
    }
    job.started = true
    completion(.success(()))
    DispatchQueue.global(qos: .userInitiated).async {
      do {
        if job.cancelled {
          job.events.sendError(code: "cancelled", message: "Cancelled")
          return
        }
        job.events.sendProgress(0.05)
        let result = try ImageCompressor.compress(
          source: source, destination: destination, options: options)
        if job.cancelled {
          job.events.sendError(code: "cancelled", message: "Cancelled")
          return
        }
        job.events.sendProgress(1)
        job.events.sendCompleted(result)
      } catch let pigeon as PigeonError {
        job.events.sendError(
          code: pigeon.code, message: pigeon.message ?? pigeon.code,
          details: pigeon.details.map { "\($0)" })
      } catch {
        job.events.sendError(code: "encode", message: error.localizedDescription, details: "\(error)")
      }
    }
  }

  func queryVideoCapabilities() throws -> VideoCapabilitiesMsg {
    VideoCompressor.queryCapabilities()
  }

  func startVideoCompress(
    id: Int64,
    inputPath: String,
    outputPath: String,
    options: VideoOptionsMsg,
    completion: @escaping (Result<Void, Error>) -> Void
  ) {
    guard let job = jobs[id] else {
      completion(
        .failure(PigeonError(code: "instanceNotFound", message: "No job \(id)", details: nil)))
      return
    }
    if job.started {
      completion(
        .failure(PigeonError(code: "invalidState", message: "Job \(id) already started", details: nil)))
      return
    }
    job.started = true
    do {
      job.video = try VideoCompressor(
        events: job.events, inputPath: inputPath, outputPath: outputPath, options: options)
      completion(.success(()))
    } catch {
      completion(.failure(error))
    }
  }

  func cancelJob(id: Int64) throws {
    guard let job = jobs[id] else {
      throw PigeonError(code: "instanceNotFound", message: "No job \(id)", details: nil)
    }
    job.cancelled = true
    job.video?.cancel()
  }

  func disposeJob(id: Int64) throws {
    if let job = jobs.removeValue(forKey: id) {
      job.cancelled = true
      job.video?.cancel()
      job.events.dispose()
    }
  }
}

private final class Job {
  let id: Int64
  let events: JobEventStream
  var started = false
  var cancelled = false
  var video: VideoCompressor?

  init(id: Int64, events: JobEventStream) {
    self.id = id
    self.events = events
  }
}
